streaming/java/streaming-runtime/src/main/java/io/ray/streaming/runtime/worker/tasks/StreamTask.java

```diff
@@ -8,51 +8,100 @@
+import io.ray.streaming.runtime.worker.context.JobWorkerContext;
 import io.ray.streaming.runtime.worker.context.StreamingRuntimeContext;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+/**
+ * {@link StreamTask} is a while-loop thread to read message, process message, and send result
+ * messages to downstream operators
+ */
 public abstract class StreamTask implements Runnable {
 
   private static final Logger LOG = LoggerFactory.getLogger(StreamTask.class);
-
-  protected int taskId;
+  private final ContextBackend checkpointState;
+  public volatile boolean isInitialState = true;
+  public long lastCheckpointId;
   protected Processor processor;
   protected JobWorker jobWorker;
   protected DataReader reader;
-  List<Collector> collectors = new ArrayList<>();
-
+  protected DataWriter writer;
   protected volatile boolean running = true;
   protected volatile boolean stopped = false;
-
+  List<Collector> collectors = new ArrayList<>();
+  private Set<Long> outdatedCheckpoints = new HashSet<>();
   private Thread thread;
 
-  protected StreamTask(int taskId, Processor processor, JobWorker jobWorker) {
-    this.taskId = taskId;
+  protected StreamTask(Processor processor, JobWorker jobWorker, long lastCheckpointId) {
     this.processor = processor;
     this.jobWorker = jobWorker;
-    prepareTask();
+    this.checkpointState = jobWorker.contextBackend;
+    this.lastCheckpointId = lastCheckpointId;
 
     this.thread = new Thread(Ray.wrapRunnable(this),
         this.getClass().getName() + "-" + System.currentTimeMillis());
     this.thread.setDaemon(true);
   }
 
+  public ChannelRecoverInfo recover(boolean isRecover) {
+
+    if (isRecover) {
+      LOG.info("Stream task begin recover.");
+    } else {
+      LOG.info("Stream task first start begin.");
+    }
+    prepareTask(isRecover);
+
+    // start runner
+    ChannelRecoverInfo recoverInfo = new ChannelRecoverInfo(new HashMap<>());
+    if (reader != null) {
+      recoverInfo = reader.getQueueRecoverInfo();
+    }
+
+    thread.setUncaughtExceptionHandler(
+        (t, e) -> LOG.error("Uncaught exception in runner thread.", e));
+    LOG.info("Start stream task: {}.", this.getClass().getSimpleName());
+    thread.start();
+
+    if (isRecover) {
+      LOG.info("Stream task recover end.");
+    } else {
+      LOG.info("Stream task first start finished.");
+    }
+
+    return recoverInfo;
+  }
+
   /**
-   * Build upstream and downstream data transmission channels according to {@link ExecutionVertex}.
+   * Load checkpoint and build upstream and downstream data transmission
+   * channels according to {@link ExecutionVertex}.
    */
-  private void prepareTask() {
-    LOG.debug("Preparing stream task.");
+  private void prepareTask(boolean isRecreate) {
```

Comment:
`isRecreate` shoule be removed if not used.

Reply:
in Java it's actually in use. it's not the is_recreate in c++